hpx/lcos/future.hpp

```diff
@@ -1499,17 +1515,32 @@ namespace hpx { namespace lcos
     >::type
     make_ready_future()
     {
-        typedef typename hpx::util::decay_unwrap<T>::type result_type;
-        typedef lcos::detail::future_data<result_type> shared_state;
-        typedef typename shared_state::init_no_addref init_no_addref;
-        typedef typename shared_state::default_construct default_construct;
+        using result_type = typename hpx::util::decay_unwrap<T>::type;
 
-        boost::intrusive_ptr<shared_state> p(
-            new shared_state(init_no_addref{}, default_construct{}),
-            false);
+        using base_allocator =
+            util::thread_allocator<detail::future_data_refcnt_base>;
+        using shared_state = typename traits::detail::shared_state_allocator<
+                lcos::detail::future_data<result_type>, base_allocator
+            >::type;
+
+        using allocator = typename std::allocator_traits<base_allocator>::
+            template rebind_alloc<shared_state>;
+        using traits = std::allocator_traits<allocator>;
+
+        using init_no_addref = typename shared_state::init_no_addref;
+        using default_construct = typename shared_state::default_construct;
+
+        using unique_ptr = std::unique_ptr<shared_state,
+            util::allocator_deleter<allocator>>;
+
+        allocator alloc{};
+        unique_ptr p(traits::allocate(alloc, 1),
+            util::allocator_deleter<allocator>{alloc});
+        traits::construct(
+            alloc, p.get(), init_no_addref{}, default_construct{}, alloc);
 
         return hpx::traits::future_access<future<result_type>>::create(
-            std::move(p));
+            p.release(), false);
```

Comment:
Is the only reason for going to the trouble of creating a unique_ptr with an  `allocator_deleter` and then releasing it right after, to ensure that if an exception is thrown, all is cleared up?

Reply:
@biddisco yes, that is correct.